File: graph/26.cpp
#include <bits/stdc++.h>

using namespace std;

using Graph = vector<vector<int>>;

bool topologicalSort(const Graph &G, vector<int> &inDegree) {
    queue<int> q;

    for (int i = 0; i < inDegree.size(); ++i) {
        if (inDegree[i] == 0) q.push(i);
    }

    while (!q.empty()) {
        auto v = q.front();
        q.pop();

        for (const auto &u : G[v]) {
            inDegree[u]--;
            if (inDegree[u] == 0) q.push(u);
        }
    }

    return accumulate(inDegree.begin(), inDegree.end(), 0) == 0;
}

void createListAdj(Graph &G, int source, int dest) { G[source].push_back(dest); }

bool canFinish(int numCourses, vector<vector<int>> &prerequisites) {
    Graph G(numCourses);
    vector<int> inDegree(numCourses, 0);

    for (const auto &edge : prerequisites) {
        createListAdj(G, edge[1], edge[0]);
        inDegree[edge[0]]++;
    }

    return topologicalSort(G, inDegree);
}

int main() {
    int numCourses = 2;
    vector<vector<int>> prerequesites = {{1, 0}, {0, 1}};

    cout << boolalpha << canFinish(numCourses, prerequesites);

    return 0;
}

File: graph/26.py
from typing import List
from collections import deque


def topological_sort(graph: List[List], in_degree: List):
    q = deque()

    for u, value in enumerate(in_degree):
        if value == 0:
            q.append(u)

    while len(q) > 0:
        v = q.popleft()

        for u in graph[v]:
            in_degree[u] -= 1

            if in_degree[u] == 0:
                q.append(u)

    return sum(in_degree) == 0


def create_adj_matrix(graph: List[List], source: int, dest: int):
    graph[source].append(dest)


def canFinish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    graph = [[] for _ in range(numCourses)]
    in_degree = [0 for _ in range(numCourses)]

    for item in prerequisites:
        create_adj_matrix(
            graph,
            item[1],
            item[0],
        )
        in_degree[item[0]] += 1

    return topological_sort(graph, in_degree)


numCourses = 5
prerequisites = [[1, 4], [2, 4], [3, 1], [3, 2]]
print(canFinish(numCourses, prerequisites))
